clevertap-core/src/main/java/com/clevertap/android/sdk/db/CtDatabase.kt

```diff
@@ -96,13 +101,63 @@ class DatabaseHelper internal constructor(val context: Context, val config: Clev
         executeStatement(db, CREATE_TEMP_USER_PROFILES_TABLE)
 
         val deviceId = getDeviceIdForAccountIdFromPrefs(config.accountId)
-        val copyUserProfilesData = """INSERT INTO temp_${USER_PROFILES.tableName} (${Column.ID}, ${Column.DEVICE_ID}, ${Column.DATA}) SELECT ${Column.ID}, '$deviceId', ${Column.DATA} FROM ${USER_PROFILES.tableName};"""
 
-        executeStatement(db, copyUserProfilesData)
+        // Query to select all data from the old user profiles table
+        val selectQuery = "SELECT ${Column.ID}, ${Column.DATA} FROM ${USER_PROFILES.tableName};"
+        val cursor = db.rawQuery(selectQuery, null)
+
+        cursor.use {
+            if (cursor.moveToFirst()) {
+                val id = cursor.getString(cursor.getColumnIndexOrThrow(Column.ID))
+                val dataString = cursor.getString(cursor.getColumnIndexOrThrow(Column.DATA))
+                val updatedDataString = migrateDataString(dataString)
+
+                // Insert the modified data into the temporary table
+                val insertQuery = """INSERT INTO temp_${USER_PROFILES.tableName} (${Column.ID}, ${Column.DEVICE_ID}, ${Column.DATA})
+                                 VALUES ('$id', '$deviceId', '$updatedDataString');"""
+                executeStatement(db, insertQuery)
+            }
+        }
+
         executeStatement(db, DROP_USER_PROFILES_TABLE)
         executeStatement(db, RENAME_USER_PROFILES_TABLE)
     }
 
+    /**
+     * This function migrates the data column for the userProfiles table
+     * Removes the "$D_" prefix from date related property values
+     * Removes the "$set" and "$add" key from incorrectly stored multi-valued properties
+     */
+    private fun migrateDataString(dataString: String): String {
+        return try {
+            val jsonObject = JSONObject(dataString)
+            val keys = jsonObject.keys()
+
+            while (keys.hasNext()) {
+                val key = keys.next()
+                var value = jsonObject.get(key)
```

Comment:
check if this throws exception or will cause NPE further in func

Reply:
if `value is String` and `value is JSoNObject ` will return false if value is null and mitigate any NPE